Add concurrent_threads_soft_limit parameter

rvasin opened the pull request:
### Changelog category:
- New Feature

### Changelog entry
Add concurrent_threads_soft_limit parameter to increase performance in case of high RPS by means of limiting total number of threads for all queries.

Closes #36551

See the attached article for details: [article.pdf](https://github.com/ClickHouse/ClickHouse/files/8707327/article.pdf)

rvasin committed:
Added global_max_threads parameter.

rvasin committed:
Corrected bug in number of threads calculation

rvasin committed:
Moved global_max_threads parameter from users.xml file to config.xml file

rvasin committed:
Added default value for globally_available_threads

rvasin committed:
Added test_global_max_threads integration test

rvasin committed:
Implemented -1 value handling for global_max_threads parameter

rvasin committed:
For value -1 of global_max_threads replaced physical cores count by logical cores count

rvasin committed:
Added thread_factor = 2 constant for initial value calculation of global_max_threads

rvasin committed:
renamed global_max_threads parameter to total_max_threads

rvasin committed:
Merge branch 'master' of github.com:ClickHouse/ClickHouse into ADQM-445

rvasin committed:
Code cleanup. Corrected Loggers path. Refactoring Context code.

rvasin committed:
Removed logging code

rvasin committed:
Moved process_list_element into if. Made more code cleanup.

rvasin committed:
Added comment about usage of total_max_threads.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=37285) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=37285) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ClickHouse/ClickHouse?pullRequest=37285) it.</sub>

rvasin commented:
Hi, can anyone comment what is the issue with CheckLabels? What should I do to correct it?

rvasin committed:
Corrected all issues found by Style Check (actions). Renamed setGlobalMaxThreads into setTotalMaxThreads.

rvasin commented:
Hi, could you please comment about status of my pull request? It seems I corrected all issues. And the remaining " 2 failing checks " are not related to the developed feature. In fact the feature is implemented such way that by default it's turned off and it does not touch any CH functionality (it does not slow don't CH etc).

azat commented on programs/server/Server.cpp:
```diff
@@ -1118,6 +1118,18 @@ int Server::main(const std::vector<std::string> & /*args*/)
             if (config->has("max_partition_size_to_drop"))
                 global_context->setMaxPartitionSizeToDrop(config->getUInt64("max_partition_size_to_drop"));
 
+            if (config->has("total_max_threads"))
```
Maybe it is better to name it `max_threads_for_all_queries` (by analogy with `max_memory_usage_for_all_queries`, now deprecated) ?
Also worth adding an example into `config.xml` (or update documentation), otherwise how to know about this new setting?
Plus in your initial article there is an explanation for this config directive, worth put it in the comment/doc.
And definitely put a word about "at least one thread will be allowed anyway"

ilejn replied:
@azat , we've failed to attract attention to the parameter name https://github.com/ClickHouse/ClickHouse/issues/36551#issuecomment-1126112304 and decided to go with _total_max_threads_
If you believe that max_threads_for_all_queries is better, we'll willingly switch to this name.

azat replied:
@ilejn thanks for the note. I don't have strong feeling about this, but `max_threads_for_all_queries` looks more "in ClickHouse style" I guess.

rvasin replied:
max_threads_for_all_queries -  it a good name. I also thought that it's more in CH-style. So I will rename it. 
The minor question is about naming the supporting functions:
```
    /// Get total number of threads for all queries in process list.
    size_t getTotalNumThreads() const;

    size_t getTotalMaxThreads() const { return total_max_threads; }

    void setTotalMaxThreads(size_t total_max_threads_)
```

What do you think should be the correct names? 

So you these these are good variants? 
```

    size_t getNumThreadsForAllQueries() const;

    size_t getMaxThreadsForAllQueries() const { return max_threads_for_all_queries; }

    void setMaxThreadsForAllQueries(size_t max_threads_for_all_queries_)
```

azat replied:
To me, functions can be leased as now, since it is already attached to processes, so simply total_max_threads is fine.

rvasin replied:
I did not understand - should I rename parameter  total_max_threads into max_threads_for_all_queries or not? (and in which places?). If we decide to rename then I would rename it everywhere to keep the same naming logic everywhere.

azat replied:
>should I rename parameter total_max_threads into max_threads_for_all_queries or not?

I think it is more in ClickHouse style, maybe you should wait until someone else will comment on this.

rvasin replied:
> Also worth adding an example into `config.xml` (or update documentation), otherwise how to know about this new setting?

@azat in the latest commit I added a documentation text into settings.md and also into config.xml please take a look. Also I added -1 as default value into config.xml to see how will CI tests will run. I will probably remove -1 and set 0 in future config.xml and remove all logging code after the pull request will be in finished form.

azat commented on programs/server/Server.cpp:
```diff
@@ -1118,6 +1118,18 @@ int Server::main(const std::vector<std::string> & /*args*/)
+                if (total_max_threads == -1)
+                {
+                    // Based on tests total_max_threads has an optimal value when it's about two times of logical CPU cores
+                    constexpr size_t thread_factor = 2;
```
This sounds small, since clickhouse uses threads for everything, for example each query has at least two threads, but they do not eats lots of CPU, so I would say that `thread_factor=10` (or maybe even 20) should be used at least by default.
What do you think?

ilejn replied:
@azat , it is usually believed that the optimal number of active threads in a system is number_of_CPU_cores, or number_of_CPU_cores*N, where N is something like 2 or 4. Our experiments proved it.
N depends on the kind of CPU (particularly if hyperthreading or similar technology is enabled) and depends on delays caused by disks subsystem and by interactions with external entities, so in some exotic cases 10 might be better than 2, but 2 or 4 seems much better default values.

ilejn replied:
Yes, if our tests were disk bound, result would be different.

azat replied:
Not only disk bound, thread may just poll something once in a ms or so.

ilejn replied:
Yes, this is true.
The worst scenario: number of disk or network bound queries that 'eat' total_max_threads, then a CPU bound query. As a result - low CPU utilization.
I think that it is not a problem of a bad thread_factor, it is just a limitation of this approach. BTW, initially, we thought about a more sophisticated way to calculate max_threads based on load average or something, but found out that total_max_threads works fine for our scenarios.
Summing up, I suggest

1. use 4, not 2 as _thread_factor_
2. make the comment in the source code more detailed, mentioning that default is good for CPU bound scenarios and maybe suboptimal for other ones
3. add an entry in the example config with long and comprehensive explanation how to estimate the proper value and when the default value is not perfect.

@azat ?

azat replied:
>make the comment in the source code more detailed, mentioning that default is good for CPU bound scenarios and maybe suboptimal for other ones

Agree, better to add some comment.

>add an entry in the example config with long and comprehensive explanation how to estimate the proper value and when the default value is not perfect.

Yes, great.

>use 4, not 2 as thread_factor

To me, 4 looks too small.

Sane defaults is very important for massive usage, since very few users looks at config.xml/documentation to tune some settings, most of them uses default, and since not all threads in ClickHouse is CPU bound, I voting for 10 at least.

But if everybody else think that 4 is good, then I'm fine either, since I will either change the default or not using this limit at all.

ilejn replied:
> To me, 4 looks too small.

> Sane defaults is very important for massive usage, since very few users looks at config.xml/documentation to tune some settings, most of them uses default, and since not all threads in ClickHouse is CPU bound, I voting for 10 at least.

At first glance it is strange to conduct massive tests (some are described in article.pdf attached to the PR), get clear results, and ignore these results afterward.

azat replied:
Ok, thinking about this a little bit more and I'm convinced me that 4 is OK as default (or even 3, since in your pdf 2-3 is optimal AFAICS), since
- this is actually not total threads limit (and hence total_max_threads does not looks a good naming), but just a threshold after what only one thread will be used for query execution.
- plus this setting does not takes into account some threads (like TCPHandler thread)

ilejn replied:
Agree with all your statements. 

I admit that total_max_threads is misleading. To me, the risk of misunderstanding should be mitigated by a comment, but maybe I am wrong and we really should think about a better name.

rvasin replied:
@azat  I pushed a new commit where I set thread_factor = 3;

azat commented on tests/integration/test_total_max_threads/test.py:
```diff
@@ -0,0 +1,49 @@
+            "select length(thread_ids) from system.query_log where current_database = currentDatabase() and type = 'QueryFinish' and query_id = 'test_total_max_threads_2'"
+        )
+        == "51\n"
+    )
```
Test does not check that limit will be decreased when the threshold will be reached. This case sounds more major then the default value.

rvasin replied:
@azat Could you please explain more in details what do you mean? Because in fact this test test_total_max_threads_defined shows that when the query requires 102 threads to execute and we have total_max_threads=50 then finally the query is executed in 52 threads. It's less than 102. So it's OK. Please explain your point of view with sample numbers.

rvasin replied:
Yes, it's an important point that in any case the query still has one thread so the in even worst case the query will be allowed to execute but in one thread. And for high CPU load conditions it's in fact the best case. It's like max_threads=1. 
Now I am working on the documentation text to clearly explain this point. 

But my question was about the integration test test_total_max_threads_defined. Do you think it's OK or how should I improve it? Because in my option the test it OK.

rvasin replied:
@azat So do you mean I should simply write a new (3rd) test for the when 1 thread will be recommended by system and finally the query will be executed in 2 threads? 
Then we could verify it by
length(thread_ids) = 2 
Do you mean that?

rvasin replied:
@azat I tried this code:

```
# total_max_threads = 10 in config.
def test_total_max_threads_reached(started_cluster):
    node4.query(
        "SELECT count(*) FROM numbers_mt(10000000)  settings max_threads=5"
    )
    node4.query(
        "SELECT count(*) FROM numbers_mt(10000000)  settings max_threads=5"
    )
    node4.query(
        "SELECT count(*) FROM numbers_mt(10000000)  settings max_threads=5", query_id="test_total_max_threads_4"
    )    
    node4.query("SYSTEM FLUSH LOGS")
    assert (
        node4.query(
            "select length(thread_ids) from system.query_log where current_database = currentDatabase() and type = 'QueryFinish' and query_id = 'test_total_max_threads_4'"
        )
        == "2\n"
    )

```
The problem is that search query is executed in sync mode. So  length(thread_ids)=7 instead of 2.

Probably I need use use some async features of CH or Python to simulate it. Do you have any suggestions?

azat replied:
You can execute queries in threads, see https://github.com/ClickHouse/ClickHouse/blob/a53cfa9fcad121830db211a5831c3eafc36fb1ac/tests/integration/test_storage_kafka/test.py#L1149-L1158 for example

rvasin replied:
@azat OK. Thank you. I will try it. I also found the following interesting apply_async usage:
https://github.com/ClickHouse/ClickHouse/blob/a53cfa9fcad121830db211a5831c3eafc36fb1ac/tests/integration/test_quorum_inserts_parallel/test.py#L46

rvasin replied:
@azat See the latest commit. I added the following test:

```
# note: total_max_threads = 10 in config_limit_reached.xml
def test_total_max_threads_limit_reached(started_cluster):
    logging.debug("ADQM: test begin")

    def thread_select():
        logging.debug("ADQM: started another thread")
        node4.query(
            "SELECT sleep(3) FROM numbers_mt(10000000) settings max_threads=100"
        )
        logging.debug("ADQM: finished another thread")

    another_thread = threading.Thread(target=thread_select)
    another_thread.start()

    time.sleep(0.5)
    logging.debug("ADQM: started main query")
    node4.query(
        "SELECT count(*) FROM numbers_mt(10000000) settings max_threads=5",
        query_id="test_total_max_threads_4",
    )
    logging.debug("ADQM: finished main query")
    another_thread.join()
    logging.debug("ADQM: logs: %s", node4.grep_in_log("ADQM"))
    node4.query("SYSTEM FLUSH LOGS")
    assert (
        node4.query(
            "select length(thread_ids) from system.query_log where current_database = currentDatabase() and type = 'QueryFinish' and query_id = 'test_total_max_threads_4'"
        )
        == "2\n"
    )
    logging.debug("ADQM: test end")
```

According to logs it seems like it works fine:

> 
> 2022.05.20 11:11:32.481001 [ 8 ] {631d3398-20b7-4c5f-ba52-1b3d7456a18b} <Debug> ADQM: max_threads: 100
> 2022.05.20 11:11:32.481003 [ 8 ] {631d3398-20b7-4c5f-ba52-1b3d7456a18b} <Debug> ADQM: Recommended num threads: 100
> 2022.05.20 11:11:32.481006 [ 8 ] {631d3398-20b7-4c5f-ba52-1b3d7456a18b} <Debug> ADQM: Total number of threads from config: 10
> 2022.05.20 11:11:32.481012 [ 8 ] {631d3398-20b7-4c5f-ba52-1b3d7456a18b} <Debug> ADQM: Current total num threads: 1
> 2022.05.20 11:11:32.481015 [ 8 ] {631d3398-20b7-4c5f-ba52-1b3d7456a18b} <Debug> ADQM: Total available threads: 9
> 2022.05.20 11:11:32.481017 [ 8 ] {631d3398-20b7-4c5f-ba52-1b3d7456a18b} <Debug> ADQM: Recommended num threads: 9
> 2022.05.20 11:11:32.481020 [ 8 ] {631d3398-20b7-4c5f-ba52-1b3d7456a18b} <Debug> ADQM: Finally num threads: 9
> 2022.05.20 11:11:33.019071 [ 9 ] {test_total_max_threads_4} <Debug> ADQM: maxParallelStreams: 5
> 2022.05.20 11:11:33.019077 [ 9 ] {test_total_max_threads_4} <Debug> ADQM: max_threads: 5
> 2022.05.20 11:11:33.019080 [ 9 ] {test_total_max_threads_4} <Debug> ADQM: Recommended num threads: 5
> 2022.05.20 11:11:33.019083 [ 9 ] {test_total_max_threads_4} <Debug> ADQM: Total number of threads from config: 10
> 2022.05.20 11:11:33.019089 [ 9 ] {test_total_max_threads_4} <Debug> ADQM: Current total num threads: 12
> 2022.05.20 11:11:33.019093 [ 9 ] {test_total_max_threads_4} <Debug> ADQM: Total available threads: 0
> 2022.05.20 11:11:33.019096 [ 9 ] {test_total_max_threads_4} <Debug> ADQM: Recommended num threads: 0
> 
> 2022.05.20 11:11:33.019099 [ 9 ] {test_total_max_threads_4} <Debug> ADQM: Finally num threads: 1

@azat Could you please comment it? Is it OK? The idea was to make one long thread working in a separate thread which consumes all query processing threads (it reaches the limit). And at the same time run our control query to verify that our algorithm gives only 1 thread to run to this query. It seems like it's working. I was in doubt about running the the long query in the same node4. But in seems like it's working: both queries runs in parallel. What do you think?

azat replied:
Yes LGTM.

rvasin committed:
Add more intergration tests. Add documenation text. Change thread_factor to 3. Add logging. Set default value = -1 to see the results CI tests.

rvasin committed:
Improve integration test test_total_max_threads_limit_reached by adding sync white loop.

rvasin commented on tests/integration/test_total_max_threads/test.py:
```diff
@@ -0,0 +1,108 @@
+    another_thread = threading.Thread(target=thread_select)
+    another_thread.start()
+
+    time.sleep(0.5)
```
@azat I pushed new commit by adding KILL QUERY  and adding wait loop. Please take a look.

azat replied:
LGTM

rvasin committed:
Style corrections

rvasin commented:
May someone comment why Fast Test failed? It's not clear from test's logs. @

azat commented:
>May someone comment why Fast Test failed? It's not clear from test's logs. @

When you have `Invalid check_status.tsv` you need to take a look at [github action check](https://github.com/ClickHouse/ClickHouse/runs/6558133989?check_suite_focus=true)

```
2022-05-23T16:14:12.9691228Z 2022-05-23 19:14:12 
2022-05-23T16:14:12.9691372Z 2022-05-23 19:14:12 All tests have finished.
2022-05-23T16:30:00.3963348Z WARNING:root:Killing process 313112, timeout 2400 exceeded
```

https://github.com/ClickHouse/ClickHouse/blob/39b37a3240f74f4871c8c1679910e065af6bea19/tests/ci/fast_test_check.py#L147

Maybe some issues with waiting processes or smth similar, not enough logs, to tell precisely.
So you can simply rebase.

rvasin commented:
> Maybe some issues with waiting processes or smth similar, not enough logs, to tell precisely. So you can simply rebase.

By rabasing do you mean the in the next commit it maybe OK?

 In fact it's the second commit where Fast Test failed. I tried to test value of total_max_threads=-1 on CI Tests. But the tests don't run because of Fast Test failure.

azat commented:
>By rabasing do you mean the in the next commit it maybe OK?

Yes.

>In fact it's the second commit where Fast Test failed

In the previous commit [`02015_global_in_threads`](https://s3.amazonaws.com/clickhouse-test-reports/37285/dc737cf6e8b9e72de5bd04b073518a880be42507/fast_test__actions_.html) fails.

azat commented:
>I tried to test value of total_max_threads=-1 on CI Tests

And that's why 02015_global_in_threads fails I guess.

rvasin commented:
> > I tried to test value of total_max_threads=-1 on CI Tests
> 
> And that's why 02015_global_in_threads fails I guess.

Azat, thank you. Yes, this test failed and it's clear why. Inside the test:
```

${CLICKHOUSE_CLIENT} --log_queries=1 --max_threads=32 --query_id "2015_${CLICKHOUSE_DATABASE}_query" -q "select count() from remote('127.0.0.{2,3}', numbers(10)) where number global in (select number % 5 from numbers_mt(1000000))"
${CLICKHOUSE_CLIENT} -q "system flush logs"
${CLICKHOUSE_CLIENT} -q "select length(thread_ids) >= 32 from system.query_log where event_date >= yesterday() and query_id = '2015_${CLICKHOUSE_DATABASE}_query' and type = 'QueryFinish' and current_database = currentDatabase()"

```

So if we have, for example, 8 logical cores and in config we have total_max_threads=-1 then 
it will calculated the limit as 8*3 = 24. But the test expects to get >=32 threads. So it will fail.
In this case probably as I wrote in documentation the good solution is to use default value for total_max_threads=0 Thus all tests will pass in this case. I will try to think maybe there are better solutions.

rvasin commented:
> > I tried to test value of total_max_threads=-1 on CI Tests
> 
> And that's why 02015_global_in_threads fails I guess.

Azat, thank you. Yes, this test failed and it's clear why. Inside the test:
```

${CLICKHOUSE_CLIENT} --log_queries=1 --max_threads=32 --query_id "2015_${CLICKHOUSE_DATABASE}_query" -q "select count() from remote('127.0.0.{2,3}', numbers(10)) where number global in (select number % 5 from numbers_mt(1000000))"
${CLICKHOUSE_CLIENT} -q "system flush logs"
${CLICKHOUSE_CLIENT} -q "select length(thread_ids) >= 32 from system.query_log where event_date >= yesterday() and query_id = '2015_${CLICKHOUSE_DATABASE}_query' and type = 'QueryFinish' and current_database = currentDatabase()"

```

So if we have, for example, 8 logical cores and in config we have total_max_threads=-1 then 
it will calculated the limit as 8*3 = 24. But the test expects to get >=32 threads. So it will fail.
In this case probably as I wrote in documentation the good solution is to use default value for total_max_threads=0 Thus all tests will pass in this case. I will try to think maybe there are better solutions.

rvasin committed:
Change max_threads=6 in 02015_global_in_threads test

rvasin commented:
I set max_threads=6 in 02015_global_in_threads test to see how other CI tests will run when total_max_threads=-1 in config.

rvasin committed:
Remove logging code

ilejn commented:
> I set max_threads=6 in 02015_global_in_threads test to see how other CI tests will run when total_max_threads=-1 in config.

As far as I can see it would work if number_of_cpu_cores*3>=6, IOW test fails if run on a single core box.
Honestly I have no idea how to keep the test meaningful and reliable, since the whole idea of '-1' is flexibility.

@rvasin Is it feasible to ignore total_max_threads if max_threads is set explicitly? This behaviour might be more natural for users. And (as a bonus) more test-friendly.
What do you think?

rvasin committed:
Set default value to 0 for total_max_threads parameter. Code style cleanup.

serxa commented:
There is a concern about current implementation. It's not work-conserving because limit is enforced on query start and in fact is applied through whole lifespan of a query. But if at some instant there is an opportunity to use more cpu released by other threads this implementation will do nothing about it.

And it will actually be a real problem on high-loaded clusters. When server operates near that boundary, this query concurrency control scheme will give almost random amount of threads to queries with high `max_threads`. Let me explain. Suppose there is mixed load: most of queries are light and do not require a lot of threads, but there are occasional heavy queries that require a lot of threads. And in production usually load is random, so performance of heavy queries will be unstable in the long run, depending on exactly one random point in past.

Actually, there is simple way to make this solution work-conserving. We have to move this limitation to lower level and allow upscaling of a query after it starts. It's easier to implement, that describe it here. Let me push a few commits to this branch.

rvasin commented:
@serxa The idea was to make a first step to improve the performance of CH in mixed load. 
So the basic idea for the first step was to use query start to calculate needed number of threads.
The solution is not ideal. But it's working. It’s kind of a good approximation.
I thought the second step (for the next pull request) could be to use some cpu monitoring/load to make the algorithm even more precise/effective.
I spoke with the management: it’s not possible to allow any external contributor to  commit to  arenadata:ADQM-445
So maybe you may create own branch on ClickHouse based on my branch arenadata:ADQM-445 and commit in your branch. Or you may just put your code here in the discussion. I will commit into my branch. You are welcome to any improvements. 
Right now I am more concerted about stability of my 4th test: it seems like it hangs on while loop. I have another variant to use the following code:

```
    def queryStarted():
        node4.query("SYSTEM FLUSH LOGS")
        return node4.query(
            "select count(*) from system.query_log where current_database = currentDatabase() and type = 'QueryStart' and query_id = 'background_query'"
        )         == "1\n"
        
         
        
    while not queryStarted():
        time.sleep(0.5)
```

rvasin commented:
3rd approach for 4th test is to use threading events like in this code below. It also works. So I not sure which variant is better.

```
def test_total_max_threads_limit_reached(started_cluster):
    background_query_started = threading.Event()
    def thread_select():
        background_query_started.set()
        node4.query(
            "SELECT count(*)  FROM numbers_mt(1e11)  settings max_threads=100",
            query_id="background_query",
        )

    another_thread = threading.Thread(target=thread_select)
    another_thread.start()
    
    while not background_query_started.is_set():
        time.sleep(0.1)
        
    node4.query(
        "SELECT count(*) FROM numbers_mt(10000000) settings max_threads=5",
        query_id="test_total_max_threads_4",
    )
    another_thread.join()
    node4.query("SYSTEM FLUSH LOGS")
    assert (
        node4.query(
            "select length(thread_ids) from system.query_log where current_database = currentDatabase() and type = 'QueryFinish' and query_id = 'test_total_max_threads_4'"
        )
        == "2\n"
    )
    node4.query("KILL QUERY WHERE user = 'default' SYNC")
```

rvasin committed:
Improve integration test test_total_max_threads_limit_reached

rvasin commented:
@serxa Yesterday I fine-tuned 4th integration test and  pushed the changes to my branch. Now it’s OK. Only 4 tests were failed. After review I found that the failed tests are not related to the developed feature.
I want to explain my point of view with concrete numbers. 
For example, we have 10 analytical queries requiring 100 threads each. And these 10 queries run simultaneously. 
With current default settings CH will create about 10*100 = 1000 threads to run these 10 queries.
And if we have for example server with 16 logical cores. It’s not good to run 1000 query processing threads.
Right now there is no way to limit it in current version of CH. 
My research on this subject (different datasets, different CPU cores count) shows that ideally we need to use factor of 2-4 to calculate total limit. For 16 logical cores the total number of threads would be 32 – 64. But not 1000.  Note: we are talking only about query processing tests not about all threads used by CH.
Thus in my feature even the simple threads management on query start gives about 13% load improvement on mixed load. And this solution works. 
What I think that if we implement more intellectual thread management (dynamic threads allocation/management on query run-time) it would give an additional 2-3% load improvement. But not 10-15% additional improvement. Thus in total 13 + 2 = 15% load improvement. It would not radically improve CH performance. While it’s still better/more precise management.
So I think the current solution is OK to be merged with master. All other changes maybe submitted with pr-improvement tag in new pull requests. 
I also want to note that the total_max_thread parameter as implemented right now is an optional parameter. It does not touch any current performance of existing servers running in production. But if clients want to improve the performance they may try to turn on this parameter. At least even manual setting by client total_max_threads = 500 for 16 logical cores server would still improve somehow the performance (4-5%) instead of un-managed 1000 or maybe 2000 threads.

serxa committed:
add concurrency control in pipeline executor

serxa commented:
@rvasin I understand your point and totally agree that having 1000 threads is not very good idea, but I anyways was going to implement concurrency control logic. And I became worried about ending up with two different approaches to the problem and two different options to enable them. Let's have just one if possible or think a little bit more if not.

I implemented low-level approach [here](https://github.com/ClickHouse/ClickHouse/pull/37558), but for now it is not configurable and obviously not well tested, but do semantics that you described, exactly. However, there is a caveat. Integration with `PipelineExecutor` is also quite simple for now and to allow upscale of a query it still allocates and constructs data structures required for all `max_threads` queries. I hope this overhead would not be very big if it runs only 1 thread. But let's check.

You have excellent research on the topic and already have tests and integration with config. Could you please merge my commit into your branch and check if it is better/worser. To avoid wasting your time, you could wait for me to add couple tests and check if my solution works as appropriate. Probably I'll do it later this week.

serxa committed:
fix comments

rvasin commented:
@serxa I quickly reviewed your code. And will study it in details tomorrow. So you may write your tests as you said. Could you please explain how do you see the current situation:
1. finally we should use combination of both approaches:
	a. in QueryPipelineBuilder::getNumThreads() we use my code to calculate number of threads considering total_max_threads parameter. 
	b. then your ConcurrencyControl/PipelineExecutor simply more intelligently manages threads. 

2. or finally you think like your low level approach is better. And my approach does not make a sense? So you will end up with using only your approach. In this case I don’t understand in your code how do you manage total_max_threads limit. So how do you solve “1000 threads” (too many threads) problem. 
Sergei, please explain it more in details.

serxa commented:
Option 2. I think we just call `ConcurrencyControl::instance().setMaxConcurrency(total_max_threads)` and use this `ConcurrencyControl` singleton for all queries. Currently `getNumThreads()` is about high level constraints of a query itself and I think it is good to keep things separated: `getNumThreads()` should represent how many threads query wants to have ideally (independent of total server load). And `ConcurrencyControl` will take care about CPU resources:
1. How many threads are actually available depending on load. Load can change over time, so upscaling is done dynamically during query execution (already done, see where `spawnThreads()` is called). Downscaling is also an option, but harder to implemented and actually wont help with performance degradation, so can be left for later implementation.
2. How resources have to be distributed over concurrent queries. For now simple round-robin scheduling is implemented (see `ConcurrentControl::schedule()`) but later we can improve it in many ways (e.g. add better support for query priorities).

Yes, approach looks complicated. I'll write general explanation in corresponding PR.

serxa commented:
@rvasin Yes, let's do 2-5 in your PR after I commit my PR with some tests added. I wrote an [explanation](https://github.com/ClickHouse/ClickHouse/pull/37558#issuecomment-1138807889) of how ConcurrencyControl works.

rvasin commented:
@serxa OK. Thank you. Lets do the way you said. 
You wrote the good explanation with charts. It's clear now.
BTW do you have any web resources/help explaining how does AsyncPipelineExecutor work? Maybe also with charts/diagrams.

serxa committed:
add tests; fix bugs; make Allocation and ConcurrencyControl classes less tighly bounded

mergify[bot] committed:
Merge branch 'master' into concurrency-control

serxa committed:
fix typos

serxa committed:
Merge branch 'concurrency-control' of github.com:ClickHouse/ClickHouse into concurrency-control

rvasin committed:
Merge remote-tracking branch 'origin/concurrency-control' into ADQM-445

rvasin committed:
Remove total_max_threads logic from QueryPipelineBuilder; Add setMaxConcurrency(total_max_threads) in Server.cpp

rvasin commented:
@serxa I merged recent changes from your branch to my branch. Please take a look. 
I am going to make some performance tests. But right now: it works fine if I run CH without any parameters:
./clickhouse-server
but if I set my config file (nothing special in it. I comment most of the parts of the config. So it's like empty config):
./clickhouse-server --config-file /etc/clickhouse-server/config.xml

I am getting many error messages like:

> 2022.05.31 19:29:09.188936 [ 11937 ] {291cd117-9b7d-41e0-9bc3-f6cf67ec722b::202205_1_3_1} <Debug> MergeTreeSequentialSource: Reading 2 marks from part 202205_2_2_0, total 8 rows starting from the beginning of the part
> 2022.05.31 19:29:09.189235 [ 11923 ] {} <Trace> BaseDaemon: Received signal 11
> 2022.05.31 19:29:09.189574 [ 11923 ] {} <Trace> BaseDaemon: Received signal 6
> 2022.05.31 19:29:09.189639 [ 12105 ] {} <Fatal> BaseDaemon: ########################################
> 2022.05.31 19:29:09.189684 [ 12105 ] {} <Fatal> BaseDaemon: (version 22.6.1.1, build id: 290460D1D55A0C5A) (from thread 11983) (no query) Received signal Segmentation fault (11)
> 2022.05.31 19:29:09.189704 [ 12105 ] {} <Fatal> BaseDaemon: Address: NULL pointer. Access: read. Unknown si_code.
> 2022.05.31 19:29:09.189730 [ 12105 ] {} <Fatal> BaseDaemon: Stack trace: 0x149b2b5a 0x14a6a034 0x14a7e558 0x14cac4b5 0x14cac1ba 0x14e8ff96 0x14cc433a 0x14cbae0a 0x14cbc34f 0xb0848fa 0xb086c0e 0x7f0c0722a947 0x7f0c072baa44
> 2022.05.31 19:29:09.189731 [ 12106 ] {} <Fatal> BaseDaemon: ########################################
> 2022.05.31 19:29:09.189795 [ 12106 ] {} <Fatal> BaseDaemon: (version 22.6.1.1, build id: 290460D1D55A0C5A) (from thread 11938) (query_id: 5bdc0b89-402d-4e28-9894-dc144796af7d::202205_9756_9760_2) (query: ) Received signal Aborted (6)
> 
> ....
> 
> 

> 2022.05.31 19:29:09.239542 [ 11922 ] {} <Information> Application: Listening for http://[::]:8123
> 2022.05.31 19:29:09.239617 [ 11922 ] {} <Information> Application: Listening for native protocol (tcp): [::]:9000
> 2022.05.31 19:29:09.239682 [ 11922 ] {} <Information> Application: Listening for replica communication (interserver): http://[::]:9009
> 2022.05.31 19:29:09.239703 [ 11922 ] {} <Information> Application: Ready for connections.
> 2022.05.31 19:29:09.275329 [ 12106 ] {} <Fatal> BaseDaemon: 9.1. inlined from ./build/../contrib/libcxx/include/__memory/unique_ptr.h:54: std::__1::default_delete<DB::PullingPipelineExecutor>::operator()(DB::PullingPipelineExecutor*) const
> 2022.05.31 19:29:09.275379 [ 12106 ] {} <Fatal> BaseDaemon: 9.2. inlined from ../contrib/libcxx/include/__memory/unique_ptr.h:315: std::__1::unique_ptr<DB::PullingPipelineExecutor, std::__1::default_delete<DB::PullingPipelineExecutor> >::reset(DB::PullingPipelineExecutor*)
> 2022.05.31 19:29:09.275393 [ 12106 ] {} <Fatal> BaseDaemon: 9.3. inlined from ../contrib/libcxx/include/__memory/unique_ptr.h:269: ~unique_ptr
> 2022.05.31 19:29:09.275405 [ 12106 ] {} <Fatal> BaseDaemon: 9. ../src/Storages/MergeTree/MergeTask.h:122: DB::MergeTask::GlobalRuntimeContext::~GlobalRuntimeContext() @ 0x1496281a in /home/roman/test/ClickHouse/build/programs/clickhouse
> 2022.05.31 19:29:09.277286 [ 12107 ] {} <Fatal> BaseDaemon: 8. ./build/../src/Storages/MergeTree/MergeTask.cpp:331: DB::MergeTask::ExecuteAndFinalizeHorizontalPart::executeImpl() @ 0x14959fd4 in /home/roman/test/ClickHouse/build/programs/clickhouse

Could you take a look? 
Also when I run any integration test it shows:

> 
> ImportError while loading conftest '/ClickHouse/tests/integration/conftest.py'.
> conftest.py:1: in <module>
>     from helpers.cluster import run_and_check
> helpers/cluster.py:23: in <module>
>     import meilisearch
> E   ModuleNotFoundError: No module named 'meilisearch'

It's strange because meilisearch is installed via pip. And I can run my Python files when I use  "import meilisearch" on my computer.

rvasin commented:
@serxa Well. After some tests I still see the "Fatal" messages in PipelineExecutor and CH drops. I used CH client: queries like show databases; use system; etc work fine. But it drops on queries with numbers_mt. 
Steps-to-re-produce:

1. Build my branch: cmake .. && ninja
2. Remove current /var/lib/clickhouse (probably it's an optional step).
3. Put config.xml and users.xml into /etc/clickhouse-server/
4. Start CH: ./clickhouse-server --config-file /etc/clickhouse-server/config.xml
5. Start CH client: ./clickhouse-client
6. Then in CH client:

> vasin :) SELECT count(*) FROM numbers_mt(10000000)
> 
> SELECT count(*)
> FROM numbers_mt(10000000)
> 
> Query id: 9dabdc39-67f2-4163-b6a7-2ae0cf210d90
> 
> ┌──count()─┐
> │ 10000000 │
> └──────────┘
> 
> 1 row in set. Elapsed: 0.008 sec. Processed 10.00 million rows, 80.00 MB (1.30 billion rows/s., 10.40 GB/s.)
> 
> vasin :) select length(thread_ids) from system.query_log where current_database = currentDatabase() and type = 'QueryFinish' and query_id = '9dabdc39-67f2-4163-b6a7-2ae0cf210d90'
> 
> SELECT length(thread_ids)
> FROM system.query_log
> WHERE (current_database = currentDatabase()) AND (type = 'QueryFinish') AND (query_id = '9dabdc39-67f2-4163-b6a7-2ae0cf210d90')
> 
> Query id: 91fedc3f-e01c-4a00-b1d8-82ac2930abab
> 
> Connecting to localhost:9000 as user default.
> Exception on client:
> Code: 210. DB::NetException: Connection refused (localhost:9000). (NETWORK_ERROR)
> 
> Connecting to localhost:9000 as user default.
> Code: 210. DB::NetException: Connection refused (localhost:9000). (NETWORK_ERROR)

In the terminal running CH server:

> 2022.06.01 15:28:53.000493 [ 41970 ] {} <Trace> BaseDaemon: Received signal 11
> 2022.06.01 15:28:53.000872 [ 42185 ] {} <Fatal> BaseDaemon: ########################################
> 2022.06.01 15:28:53.000942 [ 42185 ] {} <Fatal> BaseDaemon: (version 22.6.1.1, build id: 15AA7D1DD9FAA414) (from thread 42020) (no query) Received signal Segmentation fault (11)
> 2022.06.01 15:28:53.000971 [ 42185 ] {} <Fatal> BaseDaemon: Address: 0xe0 Access: write. Address not mapped to object.
> 2022.06.01 15:28:53.001001 [ 42185 ] {} <Fatal> BaseDaemon: Stack trace: 0x14cc4341 0x14cbae0a 0x14cbaa20 0x141f2e70 0x141f1fc3 0x13371bb9 0xb0848fa 0xb086c0e 0x7fd6ae980947 0x7fd6aea10a44
> 2022.06.01 15:28:53.007879 [ 42185 ] {} <Fatal> BaseDaemon: 3. ./build/../src/Processors/Executors/ExecutionThreadContext.cpp:70: DB::ExecutionThreadContext::executeTask() @ 0x14cc4341 in /home/roman/test/ClickHouse/build/programs/clickhouse
> 2022.06.01 15:28:53.018298 [ 42185 ] {} <Fatal> BaseDaemon: 4. ./build/../src/Processors/Executors/PipelineExecutor.cpp:216: DB::PipelineExecutor::executeStepImpl(unsigned long, std::__1::atomic<bool>*) @ 0x14cbae0a in /home/roman/test/ClickHouse/build/programs/clickhouse
> 2022.06.01 15:28:53.026697 [ 42185 ] {} <Fatal> BaseDaemon: 5.1. inlined from ./build/../contrib/libcxx/include/atomic:952: bool std::__1::__cxx_atomic_load<bool>(std::__1::__cxx_atomic_base_impl<bool> const*, std::__1::memory_order)
And CH server was shut down because of the error, 
As you see there is something with PipelineExecutor / ExecutionThreadContext - Segmentation fault,

See the attached config.xml and users.xml in configs.zip

[configs.zip](https://github.com/ClickHouse/ClickHouse/files/8814363/configs.zip)

serxa committed:
perf/safety/docs improvements

rvasin committed:
Merge remote-tracking branch 'origin/concurrency-control' into ADQM-445

rvasin committed:
Fix Fatal errors caused by using spawnThreads()

rvasin committed:
Fix Fatal errors caused by using spawnThreads()

rvasin commented:
@serxa It seems like I corrected "Fatal" error messages. We should simply call spawnThreads(); under these conditions:
```
if (!tasks.isFinished() && !checkTimeLimitSoft())
{
    // Upscale if possible
    spawnThreads();
}
```
I pushed the change to my branch.
I tried to use clickhouse-benchmark to create some load on CH server. It seems like CH server is stable now.
On Monday I will try to do some performance tests.

serxa commented:
I've got a lot of errors on server start if build with thread-sanitizer. I will check if they go away with you fix. Nice catch!

rvasin commented:
@serxa I think after applying the fix it should work fine for builds with enabled  thread sanitizer.
Could you show the complete list of parameters to use with cmake to create your build with enabled  thread sanitizer? Maybe I will also create such build and test it.

serxa commented:
@rvasin Just `cmake -DSANITIZE=thread` and everything other being the same

serxa commented:
@rvasin Looks like your fix disables spawning new threads because of `checkTimeLimitSoft()` always returning `true` normally. If I remove `!` in condition  fatal errors return back unfortunately

serxa committed:
fix single-thread mode for pulling and pushing executors to work properly

rvasin committed:
Merge remote-tracking branch 'origin/concurrency-control' into HEAD

rvasin committed:
Merge remote-tracking branch 'origin/concurrency-control' into ADQM-445

rvasin commented:
@serxa I merged changed from your branch into my branch. There is the following problem: 
I define total_max_threads in config.xml, start CH server, then start CH client. I can do commands like select / show in CH client. There are no error messages in CH server, it works fine.  But when I type exit command it does not exist CH client. Like on the screenshot: 

![Screenshot from 2022-06-21 15-28-46](https://user-images.githubusercontent.com/18441691/174799905-a0855df5-c67e-437d-9175-7cd3d18ef12f.png)

I had to press Ctrl+C to finally terminate CH client. There are similar effects when using CH benchmark tool. I tried it on two machines with the same result.
I also built your branch: CH client works fine. I can exit from it. 

I think it's relating to init of your ConcurrencyControl. Currently I use the code in Server.cpp:
```
                if (total_max_threads)
                    ConcurrencyControl::instance().setMaxConcurrency(total_max_threads);
```
Do you think it's correct?

serxa commented on programs/server/Server.cpp:
```diff
@@ -1120,6 +1121,19 @@ int Server::main(const std::vector<std::string> & /*args*/)
+                    constexpr size_t thread_factor = 3;
+                    total_max_threads = std::thread::hardware_concurrency() * thread_factor;
+                }
+                if (total_max_threads)
+                    ConcurrencyControl::instance().setMaxConcurrency(total_max_threads);
```
Separate code for absent setting and zero value is required, you must pass ConcurrencyControl::Unlimited

rvasin replied:
I think you may try in your branch set somewhere ConcurrencyControl::instance().setMaxConcurrency(50);
and you will get the similar problems.

rvasin replied:
I pushed ConcurrencyControl::Unlimited "logic: in the latest commit.

rvasin committed:
Set ConcurrencyControl::Unlimited when total_max_threads 0 or undefined

serxa committed:
unittest style/perf minor fixes

mergify[bot] committed:
Merge branch 'master' into concurrency-control

mergify[bot] committed:
Merge branch 'master' into concurrency-control

serxa committed:
Merge branch 'master' into concurrency-control

serxa committed:
fix finish() condition to account only active threads

rvasin committed:
Merge branch 'concurrency-control' of github.com:ClickHouse/ClickHouse into ADQM-445

serxa committed:
review fix

rvasin committed:
Merge branch 'concurrency-control' of github.com:ClickHouse/ClickHouse into ADQM-445

rvasin committed:
Fix integration test test_total_max_threads

rvasin commented:
@serxa I merged changes from concurrency-control into my branch. Then I corrected test_total_max_threads integration test. Everything seems working. I don't see Fatal error messages anymore. I will try to do more tests later on. But now it seems like it's working. 

Sergei, could not please comment about "1 change requested ". What should I do? I remember I implemented required changes into Server.cpp

serxa commented:
> Sergei, could not please comment about "1 change requested ". What should I do?
Yeah, already done.

Let's make one final change: rename `total_max_threads` to `concurrent_threads_soft_limit` to make it more obvious semantically. Proposed earlier `max_threads_for_all_queries` looks to me not good enough, because it does not say that limit is soft.

ilejn commented:
@serxa Are you sure that your suggestion increases clarity?
Could you explain what "soft limit" means here?

serxa commented:
@ilejn Idea was to take words from settings.md and "soft" is the opposite to "not hard" here:

> The maximum number of query processing threads, excluding threads for retrieving data from remote servers, allowed to run all queries. This is not a hard limit. In case if the limit is reached the query will still get one thread to run.

But I didn't think it may clash with unix meaning of "soft". Well, that's fair enough, though I still do not like `total_max_threads`, I have no better suggestions. If nobody else have, let's leave it as is? @rvasin @azat @KochetovNicolai

ilejn commented:
@serxa actually I am Ok with concurrent_threads_soft_limit , just shared my concern.
From my point of view we have several not very good options and the choice is more or less arbitrary.

serxa commented:
Okay, let me elaborate on name a little bit more. Verbose long name is better imo because it at least make you asking question "what is soft?" and with total_max_threads you will have to explain what it is every time, because it is easy to forget what it means.

I want to avoid reference to `max_threads`, because background activities will also benefit from using ConcurrencyControl. We do not need a lot of cpu-intensive merge threads while already having a lot of queries.

Also probably later we will have some kind of hard limit, that will limit total number of concurrently running cpu-intensive (not only for queries) threads in clickhouse-server and good name for that would be concurrent_threads_hard_limit to complement soft limit.

rvasin commented:
@serxa I think concurrent_threads_soft_limit is a good name. I will rename it.

rvasin committed:
Rename total_max_threads to concurrent_threads_soft_limit

rvasin commented:
@serxa I renamed total_max_threads to concurrent_threads_soft_limit. Also updated MD docs and Changelog entry above.

rvasin committed:
Fix Style Check in test_concurrent_threads_soft_limit
